File: app/src/main/java/art/manguste/android/gamesearch/db/DBUtils.java
package art.manguste.android.gamesearch.db;

import android.app.Activity;
import android.content.Context;
import android.os.AsyncTask;
import android.util.Log;

import androidx.room.TypeConverter;

import java.lang.ref.WeakReference;
import java.util.Date;

import art.manguste.android.gamesearch.api.HttpHandler;
import art.manguste.android.gamesearch.core.Game;
import art.manguste.android.gamesearch.core.JsonParser;
import art.manguste.android.gamesearch.core.SearchType;

import static art.manguste.android.gamesearch.api.URLMaker.formURL;


// I really should use Rx or kotlin coroutines next time
public class DBUtils {

    private static final String TAG = DBUtils.class.getSimpleName();
    private GameDatabase gameDatabase;


    // Async save the game in db
    /*public static void changeFavoriteStatus(Context context, Game game) {
        Log.d(TAG, "DB: saveGameAsFavorite ");
        (new SaveAsyncTask(context, game)).execute();
    }*/

    // Async save the game in db
    public static void changeFavoriteStatus(Context context, Game game) {
        Log.d(TAG, "DB: saveGameAsFavorite ");
        (new SaveAsyncTask(context, game)).execute();
    }

    // Convert Game.class into FavoriteGame.class
    public static FavoriteGame makeFavoriteGame(Game game){
        FavoriteGame dbGame = new FavoriteGame(
                game.getId(),
                game.getName(),
                game.getGameAlias(),
                new Date(),
                game.getReleaseDate(),
                Double.valueOf(game.getRating()),
                game.getJsonString());

        return dbGame;
    }

    private static class SaveAsyncTask extends AsyncTask<Void, Void, Void> {
        private final Game game;
        private final Context context;

        private WeakReference<Activity> weakActivity;

        public SaveAsyncTask(Context context, Game game) {
            this.game = game;
            this.context = context;
        }

        @Override
        protected Void doInBackground(Void... params) {

            if (isGameInFavorite(game.getGameAlias())){
            //if (game.isFavorite()){
                GameDatabase.getInstance(context).favoriteGameDao().deleteByAlias(game.getGameAlias());
                Log.d(TAG, "DB: removed "+game.getGameAlias()+" successfully! ");
            } else {
                Game fullGameData = null;

                if (game.getJsonString() == null) {
                    // load full data
                    String urlString = formURL(SearchType.GAME, game.getGameAlias());
                    HttpHandler sh = new HttpHandler();
                    String jsonStr = sh.makeServiceCall(urlString);
                    // parse response
                    fullGameData = JsonParser.parseGameData(jsonStr, true);
                } else fullGameData = game;

                Log.d(TAG, "DB: try to save game "+fullGameData.getGameAlias());
                FavoriteGame favGame = makeFavoriteGame(fullGameData);
                GameDatabase.getInstance(context).favoriteGameDao().insert(favGame);
                Log.d(TAG, "DB: "+fullGameData.getGameAlias()+" successfully added! ");
            }
            return null;
        }

        private boolean isGameInFavorite(String gameAlias){
            boolean gameInDB = false;
            int count = GameDatabase.getInstance(context).favoriteGameDao().IsGameInFavorite(gameAlias);
            if (count > 0){
                gameInDB = true;
            }
            return gameInDB;
        }

        @Override
        protected void onPostExecute(Void agentsCount) {
        }
    }


}

File: app/src/main/java/art/manguste/android/gamesearch/threads/DBUtils.kt
package art.manguste.android.gamesearch.threads

import android.app.Activity
import android.content.Context
import android.os.AsyncTask
import android.util.Log
import art.manguste.android.gamesearch.old_api.HttpHandler
import art.manguste.android.gamesearch.old_api.JsonParser.parseGameData
import art.manguste.android.gamesearch.old_api.URLMaker.formURL
import art.manguste.android.gamesearch.core.OLD_Game
import art.manguste.android.gamesearch.core.SearchType

import java.lang.ref.WeakReference

// I really should use Rx or kotlin coroutines next time
class DBUtils {
    /*
    private val gameDatabase: GameDatabase? = null

    private class SaveAsyncTask(private val context: Context?, private val OLDGame: OLD_Game) : AsyncTask<Void?, Void?, Void?>() {
        private val weakActivity: WeakReference<Activity>? = null


        override fun doInBackground(vararg params: Void?): Void? {
            if (isGameInFavorite(OLDGame.alias)) {
                //if (game.isFavorite()){
                context?.let { GameDatabase.Companion.getInstance(it) }!!.favoriteGameDao()!!.deleteByAlias(OLDGame.alias)
                Log.d(TAG, "DB: removed " + OLDGame.alias + " successfully! ")
            } else {
                var fullOLDGameData: OLD_Game? = null
                fullOLDGameData = if (OLDGame.jsonString == null) {
                    // load full data
                    val urlString = formURL(SearchType.GAME, OLDGame.alias)
                    val sh = HttpHandler()
                    val jsonStr = sh.makeServiceCall(urlString)
                    // parse response
                    parseGameData(jsonStr, true)
                } else OLDGame
                Log.d(TAG, "DB: try to save game " + fullOLDGameData?.alias)
                val favGame = makeFavoriteGame(fullOLDGameData!!)
                context?.let { GameDatabase.Companion.getInstance(it) }!!.favoriteGameDao()?.insert(favGame)
                Log.d(TAG, "DB: " + fullOLDGameData.alias + " successfully added! ")
            }
            return null
        }

        private fun isGameInFavorite(gameAlias: String?): Boolean {
            var gameInDB = false
            val count: Int = context?.let { GameDatabase.Companion.getInstance(it) }!!.favoriteGameDao()!!.IsGameInFavorite(gameAlias)
            if (count > 0) {
                gameInDB = true
            }
            return gameInDB
        }

        override fun onPostExecute(agentsCount: Void?) {}

        companion object {
            // TODO
            fun makeFavoriteGame(OLDGame: OLD_Game): FavoriteGame {
                TODO("Not yet implemented")
                // TODO fix
             *//*    return FavoriteGame(
                        game.idgetId(),
                        game.getName(),
                        game.getGameAlias(),
                        Date(),
                        game.getReleaseDate(),
                        Double.valueOf(game.getRating()),
                        game.getJsonString())*//*
            }
        }


    }

    companion object {
        private val TAG = DBUtils::class.java.simpleName

        // Async save the game in db
        fun changeFavoriteStatus(context: Context?, OLDGame: OLD_Game) {
            Log.d(TAG, "DB: saveGameAsFavorite ")
            SaveAsyncTask(context, OLDGame).execute()
        }
    }*/
}